### PR TITLE
test: add accessibility checks for documentation examples

### PR DESCRIPTION
**Pull request by adrianschmidt**
> **Stacked on #4101** (the example-tests harness). Base is `test/example-scenario-tests`; review #4101 first. I'll retarget to `main` once that merges.

## What

Runs **axe-core** against every documented example, loaded through the docs' `#/debug/<tag>` route (reusing the runtime-tests harness). Adds one spec (`example-tests/accessibility.spec.ts`) and a committed baseline (`example-tests/axe-baseline.json`), plus a separate non-blocking **Accessibility Tests** CI job.

## Scoping — we test the *production* components, not the demo plumbing

For each example the scan is scoped to the example's own top-level content, **excluding**:
- the Kompendium chrome (we resolve the example element via a shadow-piercing locator and run `axe.run` on its nodes), and
- every `limel-example-*` scaffolding component (`limel-example-controls`, `limel-example-value`, `limel-example-event-printer`, …) wherever it appears in the subtree.

So axe measures the real components a reviewer cares about, not the example wrapper or its control panels.

## Baseline (ratchet down)

The library has substantial pre-existing a11y debt — **303 / 440 examples** currently violate at least one rule (top offenders: `color-contrast`, `aria-required-children/parent`, `label`). Making axe "must be zero" now would be permanently red, so:

- `axe-baseline.json` records the rule IDs each example currently violates.
- The test **fails only when an example violates a rule not in its baseline** — new regressions are caught, existing debt is tolerated.
- Fixing an example and deleting its baseline entry locks in the gain. New examples start with no baseline, so they must be clean.
- Regenerate after fixing issues: `UPDATE_AXE_BASELINE=1 npx playwright test --config playwright.example-tests.config.ts accessibility`. Regeneration only ratchets **down** — it removes rules that are no longer violated but never adds new ones, so rerunning it can't silently accept a regression (a genuinely new violation must be fixed, or added to the baseline by hand).

**How precise is the baseline?** It records *which rules* each example breaks — not how many times, or on which elements. So if `limel-example-foo` already has a contrast problem, its entry is simply `color-contrast`.

- ✅ The test **fails** if an example starts breaking a **rule it didn't break before**.
- ⚠️ It **won't fail** if an example gains *more instances* of a rule it already breaks (e.g. a second low-contrast element).

Tracking rules rather than individual elements keeps the baseline small and stable: it doesn't churn every time the markup shifts around. The blind spot above is the price of that — a deliberate trade to avoid a brittle, high-maintenance baseline.

## Determinism

The scan waits for async content (data fetches, editors, tables) to settle before running, so results are the stable end state. Verified deterministic: repeated full accessibility runs (440 examples) are green.

## Notes
- Non-blocking, same as the runtime suite — bake, then promote.
- Runtime and accessibility are separate CI jobs; each currently builds the docs independently. A small follow-up PR will dedupe that via a shared docs artifact and shard the accessibility run to cut wall-clock.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Automated accessibility checks added to the PR workflow

* **Tests**
  * Accessibility scanning runs against all example demos
  * Compares results to a persisted baseline and flags new violations
  * Supports updating the baseline when desired
  * Enforces a minimum example coverage guard

* **Chores**
  * Added accessibility test script and axe-core dev dependency

### CODE REVIEW

**Comment by adrianschmidt** [comment]
# Consolidated PR Review

*Multi-agent review (7 dimensions) via Claude Code.*

## PR Summary

PR #4102 adds axe-core accessibility scanning of all 440 documented examples via the docs' `#/debug/<tag>` route, reusing the #4101 example-tests harness. It adds one spec (`example-tests/accessibility.spec.ts`), a committed ratchet-down baseline (`axe-baseline.json`, the bulk of the +1428 lines), a non-blocking `accessibility-tests` CI job, a pinned `axe-core@4.12.0` devDependency, and a `resolveWorkers()` tweak forcing single-worker baseline regeneration. Base is `test/example-scenario-tests` (stacked on #4101).

## Merge Readiness -- BLOCKED ❌

The PR is a near-strict improvement over the target branch — published package surface untouched, ratchet logic correct, scoping design sound — but it introduces one operational regression: a new 440-test CI job with no hang backstop, in a workflow whose directly-analogous sibling job guards exactly that scenario. The fix is one line.

- **Blockers:**
  1. **[High from Performance]** The new `accessibility-tests` job (`.github/workflows/pr-checks.yml:95-117`) omits `timeout-minutes`, inheriting GitHub's 360-minute default. The sibling `runtime-tests` job sets `timeout-minutes: 30` with an explicit comment that a mass failure (each test retried once at 30s) "could otherwise pin the runner for a long time" — this job has the same 440-test/retry profile *plus* a slower per-test axe scan. A mass-hang failure pins a runner for ~4–6 hours per push until killed. The sibling asymmetry documents the team's own intent; the omission reads as accidental. Add `timeout-minutes: 30` (or slightly higher for axe overhead).
- **Reclassified, not a blocker:** The Observability agent rated the missing trace-upload step `[High]`. Reclassified to `[Medium]`: failures still carry the example tag + violated rule IDs and a documented local repro path, so the new job remains strictly more diagnostic capability than the target branch (which has no accessibility signal at all). It's a sibling-parity gap worth fixing — ideally in the same one-line-fix pass as the blocker — but not a regression.
- **Non-blocking but worth addressing:** See Top Recommendations — notably the trace upload, the `--workers`-override baseline-corruption footgun, and the spec-level duplication/missing count guard.

---

## 1. Backward Compatibility -- SAFE ✅

Genuinely test-only with respect to the published package: `files` is `["dist/", "index.scss", "CHANGELOG.md"]`, axe-core is a devDependency, no exports/props/events touched, and the new CI job is correctly absent from `automerge.needs`.

- **[Low]** Stacked-PR retarget risk: the `resolveWorkers()` edit modifies a line #4101 introduced, and the job assumes #4101's harness (`serve-www.cjs`, `sirv`, base config). Merging out of order or retargeting to `main` before #4101 lands would not stand alone. Author-owned and explicitly called out in the PR description.
- **Positives:** `resolveWorkers()` is a careful superset — byte-for-byte identical behavior for the existing runtime suite in every real scenario (CI `'50%'`, local `undefined`).

## 2. Code Quality -- GOOD ✅

The ratchet logic is correct on the documented path; the issues are robustness gaps around it.

- **[Medium]** `accessibility.spec.ts:114-141` — the single-writer invariant for the `afterAll` baseline writer is only *configured* (`resolveWorkers()` → 1), not *enforced*. A `--workers=N` CLI override silently defeats it: each worker writes its own partial `generated` map and the last writer clobbers `axe-baseline.json` — silent corruption of the exact guarantee the ratchet design rests on. Cheap fix: assert `test.info().config.workers === 1` in update mode and fail loudly.
- **[Medium]** `accessibility.spec.ts:90-92, 97-103` — `JSON.parse(readFileSync(...))` of the baseline has no error handling; a corrupt/half-written baseline (e.g. from the clobber above) throws a cryptic `SyntaxError` at module load. The kompendium parse shares the pattern but is pre-existing from #4101; the baseline parse is new.
- **[Medium]** Duplication with `runtime.spec.ts`: the kompendium-load/TAGS pipeline (`accessibility.spec.ts:96-112` vs `runtime.spec.ts:18-35`) and render-wait logic (`:147-156` vs `:73-86`) are copied verbatim — and the copy *dropped* the `> 400` count guard, so a broken enumeration silently reduces the accessibility suite to zero tests (the runtime job in the same workflow would still catch it). Extract a shared `example-tests` helper (TAGS loader + `waitForExampleRender`).
- **[Low]** Hand-rolled `AxeRunner`/`Baseline` type assertions (`:178-184`) — acceptable for test code; exact-pinned axe mitigates drift.
- **[Low]** `.catch(() => undefined)` on `networkidle` (`:161-163`) is intentional but undocumented at the call site — a one-line comment prevents a future "fix" into a hard wait.
- **[Low]** The failure message's "cannot be regenerated to accept new violations" slightly overstates a guarantee the `--workers` footgun can in fact defeat.
- **Positives:** Bootstrap-vs-ratchet branching, the still-violated ∩ previously-baselined intersection, and deterministic sorted output are carefully done and clearly commented.

## 3. Architecture -- GOOD ✅

Kompendium coupling is reused from #4101 rather than re-invented; the include/exclude scoping design is clean.

- **[Low]** The `LIMEL-EXAMPLE-` prefix exclusion (`accessibility.spec.ts:187-188`) was verified robust today (all scaffolding helpers match; no production component can), but it's an undocumented load-bearing invariant — unlike `runtime.spec.ts:5-11`, which documents its Kompendium couplings. One comment fixes it.
- **[Low]** Raw `axe.min.js` injection couples to axe-core's internal file layout (`node_modules/axe-core/axe.min.js`) — but the stated reason for bypassing `@axe-core/playwright` is verified sound: element-reference scoping across nested shadow roots genuinely can't be expressed in the wrapper's selector API.
- **[Low]** `Boolean(process.env.UPDATE_AXE_BASELINE)` treats `UPDATE_AXE_BASELINE=0` as truthy. Consistent between spec and config (no divergence bug), just a minor footgun.
- **[Low]** Spec duplication — see Code Quality `[Medium]`.
- **Positives:** The ratchet-down regeneration with its enforced single-worker coupling in config, and the per-tag rule-set baseline granularity (stability over instance-precision, with the trade-off documented honestly in the PR description), are genuinely well-designed.

## 4. Security -- GOOD ✅

No new exposure paths; supply-chain posture is good.

- **[Low]** `.github/workflows/pr-checks.yml` — new `actions/cache@v5` uses a floating major tag while every third-party action in the workflow is SHA-pinned. Pre-existing asymmetry (the `test` and `runtime-tests` jobs already do the same); fix all three occurrences in a follow-up.
- **[Low]** `tag` interpolated into URL/selector (`accessibility.spec.ts:145,147`) — build-derived, first-party input; byte-for-byte the pattern already on the target branch in `runtime.spec.ts`. Informational.
- **Positives:** Exact-pinned, integrity-hashed, dev-only axe-core; MPL-2.0 verified compatible (unmodified, never distributed); the job runs PR code with default read-only token, no secrets, no `permissions:` escalation.

## 5. Observability -- FAIR ⚠️

The new gate is meaningfully weaker on failure diagnosability than the sibling it extends.

- **[Medium]** `pr-checks.yml:95-117` — the shared config sets `trace: 'retain-on-failure'`, so the accessibility job *generates* Playwright traces on failure and then discards them: unlike `runtime-tests` (lines 87-93), it has no upload step. One copy-pasted step (artifact name `accessibility-tests-traces`) restores parity. *(Reclassified from the agent's `[High]` — see Merge Readiness.)*
- **[Medium]** `accessibility.spec.ts:209-228` — the assertion discards everything but `violation.id`; axe's `nodes[].target`/`failureSummary`/`impact` (the actually-actionable data) never reach the failure message. Including node details for just the new-violation rules would make CI failures self-contained. Compounds the missing trace upload.
- **[Low]** The `UPDATE_AXE_BASELINE` writer emits no summary of what changed — `git diff` is the only signal.
- **[Low]** Silent `networkidle` catch leaves no breadcrumb when a scan proceeds on a possibly-unsettled DOM.
- **[Low]** Missing `> 400` count guard — see Code Quality `[Medium]` (duplication).
- Missing job timeout: see Performance `[High]`.
- **Positives:** The failure message explains the ratchet semantics to whoever hits it, and the settle protocol (bounded networkidle + double-rAF) is sensible and empirically verified deterministic.

## 6. Performance -- FAIR ⚠️

No bundle impact (verified: axe-core dev-only, nothing leaks into `dist/` or the `files` allowlist). The issues are CI-operational.

- **[High]** `pr-checks.yml:95-117` — no `timeout-minutes` on a 440-test job whose sibling sets `timeout-minutes: 30` explicitly against the mass-failure runner-pin scenario. Inherits GitHub's 360-minute default; a mass-hang failure pins a runner ~4–6h per push. **Merge blocker** (one-line fix).
- **[Medium]** Full duplicate `npm ci` + browser install + `docs:build` roughly doubles the example-suite CI footprint. Knowingly deferred — the PR description commits to a shared-docs-artifact + sharding follow-up.
- **[Medium]** ~500KB `AXE_SOURCE` injected via `addScriptTag` into all 440 pages — inherent to per-test isolation without the wrapper; real but acceptable overhead.
- **[Medium]** Settle protocol can spend up to 5s × 440 (~37 min worst case, mostly hidden by parallelism). Bounded and intentional; flagged as cost, not defect.
- **[Low]** `[...root.querySelectorAll('*')]` per test — negligible next to `axe.run` itself.
- **[Low]** `npm run test:examples` (unfiltered) now runs both specs (~880 tests) locally — arguably intended; worth a docs note.
- **Note for #4101:** this PR's `resolveWorkers()` makes the runtime job's "one worker runs the examples serially" comment (`pr-checks.yml:62-63`) stale — that comment belongs to #4101 to reconcile. *(Verified: the `'50%'` CI value predates this PR, so the comment was already stale; nothing for this PR to fix.)*
- **Positives:** Worker resolution is single-threaded only where correctness requires it; the networkidle wait is properly bounded so it cannot hang a test.

## 7. Lime Platform Issues -- SAFE ✅

No LPID catalog rule applies. The changes are Playwright test harness and CI plumbing in a StencilJS component library — no limeobject hooks, endpoints, event handlers, tasks, or shipped component source are touched, and no pre-existing LPID issue exists in the changed files.

- **Positives:** The ratchet-down baseline is a sound, low-noise way to drive down accessibility debt with zero platform-runtime cost.

---

## Overall Verdicts

| Dimension | Verdict |
|---|---|
| Backward Compatibility | **SAFE** ✅ |
| Code Quality | **GOOD** ✅ |
| Architecture | **GOOD** ✅ |
| Security | **GOOD** ✅ |
| Observability | **FAIR** ⚠️ |
| Performance | **FAIR** ⚠️ |
| Lime Platform Issues | **SAFE** ✅ |
| **Merge Readiness** | **BLOCKED** ❌ |

## Top Recommendations

1. **[High from Performance]** Add `timeout-minutes: 30` (or ~45 to allow for axe overhead) to the `accessibility-tests` job in `pr-checks.yml` — mirrors the sibling job's deliberate mass-failure backstop. One line; the sole merge blocker.
2. **[Medium from Observability]** Add the trace-upload-on-failure step to the `accessibility-tests` job (copy from `runtime-tests`, artifact name `accessibility-tests-traces`) — traces are already generated and currently discarded.
3. **[Medium from Code Quality]** Enforce the single-writer invariant in update mode: fail loudly if `UPDATE_AXE_BASELINE` is set and workers > 1, so a `--workers` override can't silently corrupt the baseline.
4. **[Medium from Code Quality]** Extract the shared kompendium/TAGS loader and render-wait helper used by both specs — and carry the `> 400` count guard into the accessibility suite so a broken enumeration can't silently zero its coverage.
5. **[Medium from Code Quality]** Wrap the baseline `JSON.parse` with a contextual error ("axe-baseline.json is corrupt — regenerate with UPDATE_AXE_BASELINE=1") so the new failure mode is diagnosable.
6. **[Medium from Observability]** Surface axe node details (`target`, `failureSummary`) for new-violation rules in the assertion message so CI failures are self-contained.
7. **[Medium from Performance]** Land the planned shared-docs-artifact + sharding follow-up promptly — the duplicated docs build doubles the example-suite CI footprint today.
8. **[Low from Security]** SHA-pin all three `actions/cache@v5` occurrences in `pr-checks.yml` together in a follow-up.

The verdict hinges on one line: with `timeout-minutes` added (and ideally #2 in the same push), this flips to READY TO MERGE — everything else is follow-up material on a well-designed, net-positive PR.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

**Comment by adrianschmidt** [comment]
## Review follow-up

Worked through the recommendations. Addressed the blocker and five more in fixup `4c818cde3`; the rest deferred with reasoning. Verified locally: `docs:build` + the full accessibility suite is green (**441 passed** — 440 examples + the new count guard), the committed baseline is untouched, and lint is clean.

### ✅ Addressed

1. **`timeout-minutes: 45` on `accessibility-tests`** *(the blocker)*. Same ~440-test/retry profile as `runtime-tests`, plus a slower per-test axe scan, so a touch more headroom than its 30.
2. **Trace upload on failure** — copied the `runtime-tests` step (artifact `accessibility-tests-traces`); traces were generated by `retain-on-failure` and then discarded.
3. **Enforce the single-writer invariant.** A `beforeAll` now fails loudly if `UPDATE_AXE_BASELINE` is set with more than one worker, so a `--workers` override can't defeat the config's single-worker forcing and silently corrupt the baseline. (This also makes the failure message's "cannot be regenerated to accept new violations" claim true again.)
4. **Carried the `> 400` count guard** into this suite — the copy from `runtime.spec.ts` had dropped it, so a broken enumeration could silently zero accessibility coverage.
5. **Contextual baseline-parse error** — a corrupt `axe-baseline.json` now throws "is corrupt … regenerate with UPDATE_AXE_BASELINE=1" instead of a raw `SyntaxError`.
6. **Axe node details in failures** — the assertion now surfaces each new violation's node `target` + `failureSummary`, so a CI failure says *which element* broke the rule and *why*, not just the rule id. The baseline still compares rule ids only, so behavior is unchanged (verified: green run, baseline untouched).

Plus the two `[Low]` doc comments: the intentional `networkidle` `.catch`, and the `limel-example-` scaffolding-exclusion invariant.

### ⏭️ Deferred (with reasoning)

- **Extract a shared `example-tests` helper (rec `#4`, other half).** The duplicated TAGS loader / render-wait also lives in `runtime.spec.ts`, which belongs to #4101's commit — extracting cleanly means touching an already-reviewed PR's commit and re-reviewing it. I carried the missing count guard here (the actual correctness gap); the DRY refactor is cleaner as a dedicated follow-up once the stack lands. (Same call I made on #4101.)
- **Shared-docs-artifact + sharding (rec `#7`).** Already implemented in the stacked follow-up **#4103**.
- **SHA-pin `actions/cache@v5` (rec `#8`, `[Low]`).** Pre-existing across `test`/`runtime-tests`/`accessibility-tests`; per the review's own suggestion I'll pin all occurrences together — I'll batch it with the `actions/upload-artifact@v4` pin I already owe on #4103.
- Minor `[Low]`s: `UPDATE_AXE_BASELINE=0` truthiness (consistent between spec and config; nobody sets `=0`), hand-rolled axe types (acceptable for test code, exact-pinned), and the update-mode change summary.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

**Comment by adrianschmidt** [comment]
# Consolidated PR Review (re-review)

*Multi-agent review (7 dimensions) via Claude Code. Re-review of the PR after fixup `4c818cde3` addressed the prior consolidated review.*

## PR Summary

PR #4102 adds axe-core accessibility scanning of all ~440 documented examples via the docs' `#/debug/<tag>` route, reusing the #4101 example-tests harness: one spec (`example-tests/accessibility.spec.ts`), a committed ratchet-down baseline (`axe-baseline.json`, the bulk of +1505/−1), a non-blocking `accessibility-tests` CI job, and a pinned `axe-core@4.12.0` devDependency. The diff under review includes fixup `4c818cde3`, which addressed the prior review's blocker and five `[Medium]`s.

## Merge Readiness -- READY TO MERGE ✅

The PR is strictly better than the target branch, and every prior-review blocker and addressed item was independently verified present and correct in the fixup — including the sole prior `[High]` (`timeout-minutes`), whose arithmetic was re-checked (~440 tests + axe overhead fits comfortably under 45 min with all waits bounded). No agent found a new regression.

- **Blockers:** None. The prior `[High from Performance]` (missing job timeout) is fixed: `timeout-minutes: 45` verified at `pr-checks.yml:100`, sanely sized vs the sibling's proven 30.
- **Non-blocking but worth addressing:** See Top Recommendations — the deferred shared-helper extraction and the #4103 CI-dedupe follow-up (already up) carry the remaining `[Medium]`s; the rest are `[Low]`s.

---

## 1. Backward Compatibility -- SAFE ✅

Strictly additive: published surface (`files` = `dist/`, `index.scss`, `CHANGELOG.md`) untouched, axe-core is dev-only, and `resolveWorkers()` is behavior-preserving for the existing runtime suite (CI `'50%'`, local `undefined`, unchanged when `UPDATE_AXE_BASELINE` is unset).

- **[Low]** Stacked-PR retarget/ordering risk (unchanged from prior review): `resolveWorkers()` rewrites a line #4101 introduced, and the spec reuses #4101's harness conventions. Author-owned, disclosed in the description.
- **Positives:** No consumer-visible change of any kind; the umbrella `test:examples` picking up the new spec is the intended additive behavior.
- *Fixes verified:* timeout, trace upload, worker guard, count guard — all present.

## 2. Code Quality -- GOOD ✅

All six claimed fixes verified present **and correct** — including the subtle ones: the `beforeAll` guard's `test.info().config.workers !== 1` comparison is valid (Playwright resolves `'50%'` to an integer before `FullConfig`), and the `afterAll` ratchet logic was re-audited for partial-run/`-g`/retry edge cases with no bug found.

- **[Medium]** Duplication with `runtime.spec.ts` (deferred, acknowledged): the kompendium/TAGS loader (`accessibility.spec.ts:54-70` vs `runtime.spec.ts:18-35`) and render-wait poll are near-verbatim copies; the copy also lacks the runtime spec's "couples to three Kompendium internals" doc comment, so the accessibility spec relies on the sibling file for that context. Deferral reasoning (duplicate lives in #4101's commit; extract once the stack lands) is sound — not a blocker. *(Cross-referenced from Architecture.)*
- **[Low]** `kompendium.json` is parsed with a bare `JSON.parse` (`accessibility.spec.ts:55`) while the baseline parse got a contextual error — pre-existing pattern from #4101; a one-line wrap would restore symmetry.
- **[Low]** `UPDATE_AXE_BASELINE=0` truthiness — now read in three places, all consistent. Deferred previously; unchanged.
- **[Low]** Hand-rolled `AxeNode`/`AxeRunner` types rely silently on axe's runtime shape; a `// keep in sync with axe-core X.Y` marker would help. Deferred previously; unchanged.
- **Positives:** Excellent comment discipline (every non-obvious decision explained at point of use); the genuinely subtle ratchet/bootstrap algorithm is implemented correctly including partial-run and sharding edge cases.
- *Fixes verified:* worker guard ✓ correct; baseline-parse error ✓; count guard ✓; node details ✓ well-formatted; both doc comments ✓.

## 3. Architecture -- GOOD ✅

The suite extends the #4101 harness cleanly; load-bearing invariants (scaffolding exclusion, networkidle catch) are now documented, and the baseline single-writer model was verified to survive #4103's sharding direction (regeneration is local-only; CI shards never write).

- **[Low]** `resolveWorkers()` embeds an accessibility-spec concern (`UPDATE_AXE_BASELINE`) in the *shared* config — as more suites join (the stated visual-tests future), per-spec env coupling accrues. Fine for two specs; flagging the direction.
- **[Low]** `include.length === 0` returns `[]` silently (`accessibility.spec.ts:186-188`): an example whose top-level children are all scaffolding would scan nothing with no signal. No current example hits it; a scanned-examples sanity counter would close the blind spot.
- **[Low]** Spec duplication — see Code Quality `[Medium]`.
- **Positives:** The shadow-root include/exclude scoping was verified against real example sources and correctly isolates production components; bypassing `@axe-core/playwright` is justified (element-reference scoping across nested shadow roots isn't expressible in the wrapper's API).
- *Fixes verified:* scaffolding-invariant comment ✓ accurate; networkidle comment ✓ accurate.

## 4. Security -- GOOD ✅

No new exposure path. The new trace-upload step was specifically audited (lime-elements is public): traces capture only the locally-served docs built from the public repo — no secrets, auth, or external authenticated traffic — and the step is SHA-pinned, `if: failure()`, 7-day retention, identical to the runtime sibling.

- **[Low]** `actions/cache@v5` floating tag (3 occurrences incl. the new job) vs SHA-pinned everything else — pre-existing asymmetry; deferral (batch all pins on #4103) is reasoned. Unchanged severity.
- **[Low]** `tag` interpolated into URL/selector — build-derived first-party input, same pattern as the base branch. Informational.
- **Positives:** axe-core exact-pinned, integrity-hashed, dev-only, MPL-2.0-compatible, injected from local `node_modules` (never fetched at runtime); the eslint-ignored baseline contains only tag keys and rule IDs — a low-value tamper surface that can't leak data.
- *Fixes verified:* trace upload added without weakening posture ✓.

## 5. Observability -- GOOD ✅

(Prior verdict FAIR; both `[Medium]` gaps are fixed and verified.) A CI failure is now self-contained: rule ID + offending element targets + axe failure summary in the message, plus a trace artifact.

- **[Low]** `impact` (critical/serious/moderate/minor) is still dropped from violation nodes — `failureSummary` largely compensates; would help triage urgency.
- **[Low]** `UPDATE_AXE_BASELINE` runs emit no change summary — explicitly deferred ("git diff sufficient"); reasonable for a local maintainer workflow.
- **[Low]** No machine-readable reporter — only relevant once #4103 shards; out of scope here.
- **Positives:** Worker-guard, baseline-corruption, and count-guard errors all carry actionable contextual messages; the `describe` formatter produces well-indented per-node output.
- *Fixes verified:* trace upload ✓ exact sibling copy; node details ✓; count guard ✓; parse error ✓.

## 6. Performance -- GOOD ✅

(Prior verdict FAIR; the sole `[High]` is fixed.) Zero published-bundle impact (axe-core dev-only, nothing in `files`). Remaining `[Medium]`s are the previously-accepted or knowingly-deferred costs, unchanged in severity.

- **[Medium]** Duplicate CI footprint (`npm ci` + browsers + `docs:build` per job, `pr-checks.yml:101-119`) — deferred to #4103, which is already up. Unchanged.
- **[Medium]** ~500KB `AXE_SOURCE` injected into all ~440 pages — inherent to per-test isolation; previously accepted. Unchanged.
- **[Medium]** Settle protocol (≤5s networkidle + double-rAF × 440) — bounded, intentional; even the pathological all-5s case stays under the 45-min budget. Previously accepted. Unchanged.
- **[Low]** Unfiltered `npm run test:examples` now runs ~880 tests locally — worth a one-line docs note; the split scripts mitigate.
- **Positives:** All fixup code is off the green-run hot path (guard runs once, formatting only on failure, upload gated on `if: failure()`); no unbounded wait can silently consume the job budget.
- *Fix verified:* `timeout-minutes: 45` ✓ present, arithmetic checks out (runtime's proven 30 + ~7–22 min axe overhead inside the +15 headroom) — a healthy-but-slow run won't flake.

## 7. Lime Platform Issues -- SAFE ✅

No catalog rule applies: the changes are TypeScript Playwright test harness and CI plumbing — no limeobject hooks, endpoints, handlers, tasks, file-storage access, or shipped component source. The catalog's testing rules are Python/pytest-specific; the frontend rules target shipped Stencil component source, not specs. The fixup likewise introduces no runtime anti-pattern.

- **Positives:** Defensively designed test infrastructure (single-writer guard, count guard, failure-trace artifacts) with zero LPID exposure.

---

## Overall Verdicts

| Dimension | Verdict |
|---|---|
| Backward Compatibility | **SAFE** ✅ |
| Code Quality | **GOOD** ✅ |
| Architecture | **GOOD** ✅ |
| Security | **GOOD** ✅ |
| Observability | **GOOD** ✅ |
| Performance | **GOOD** ✅ |
| Lime Platform Issues | **SAFE** ✅ |
| **Merge Readiness** | **READY TO MERGE** ✅ |

## Top Recommendations

No `[High]` or `[Critical]` findings remain. The remaining items are follow-up material:

1. **[Medium from Code Quality]** Land the shared `example-tests` helper extraction (TAGS loader + render-wait, with the Kompendium-coupling doc comment) once the stack merges — the deferral is sound, but until then the two specs will drift independently.
2. **[Medium from Performance]** Merge #4103 promptly — it already implements the shared-docs-artifact + sharding that resolves the duplicated CI footprint.
3. **[Low from Observability]** Capture `impact` alongside `target`/`failureSummary` in violation nodes — one field, and failures become triage-ranked.
4. **[Low from Architecture]** Consider a scanned-examples counter so an example whose top-level children are all scaffolding can't silently scan nothing.
5. **[Low from Security]** SHA-pin the three `actions/cache@v5` occurrences in the batched pin pass already planned on #4103.

The prior review's condition — "with `timeout-minutes` added (and ideally the trace upload in the same push), this flips to READY TO MERGE" — is met: both are in fixup `4c818cde3`, verified present and correct by multiple independent agents, alongside four more addressed `[Medium]`s. The PR is a net improvement over its target branch with all remaining work explicitly owned.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

**Comment by adrianschmidt** [comment]
@coderabbitai review

File: example-tests/accessibility.spec.ts [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+    try {
+        return JSON.parse(raw) as Baseline;
+    } catch (error) {
+        throw new Error(
+            `${BASELINE_PATH} is corrupt and could not be parsed ` +
+                `(${(error as Error).message}). Regenerate it with ` +
+                `UPDATE_AXE_BASELINE=1.`
+        );
+    }
+};
```
[the remark this answers is not in the document]
Comment (adrianschmidt) [reply example-tests/accessibility.spec.ts:46]: @coderabbitai good catch — fixed in 780384caa. `parseBaseline` runs at module load regardless of mode, so a corrupt file throws before `UPDATE_AXE_BASELINE=1` could regenerate it. The message now tells you to delete or repair the file first; deleting routes through the bootstrap path that writes a fresh baseline.

File: .github/workflows/pr-checks.yml [hunk 2]
```diff
@@ -92,6 +92,42 @@ jobs:
...
+    - uses: actions/checkout@de0fac2e4500dabe0009e67214ff5f5447ce83dd # v6.0.2
+    - uses: ./.github/actions/set-up-node
+    - run: npm ci
+    - name: Cache Playwright browsers
+      id: playwright-cache
+      uses: actions/cache@v5
+      with:
+        path: ~/.cache/ms-playwright
+        key: playwright-${{ runner.os }}-${{ hashFiles('package-lock.json') }}
+        restore-keys: playwright-${{ runner.os }}-
```
[the remark this answers is not in the document]
Comment (adrianschmidt) [reply .github/workflows/pr-checks.yml:111]: @coderabbitai agreed. This is handled in the stacked follow-up #4103 (commit 97bbb8dea), which SHA-pins `actions/cache` to `27d5ce7` (v5.0.5) across the `test`, `runtime-tests` and `accessibility-tests` jobs — matching the `upload-artifact@v7.0.1` / `download-artifact@v8.0.1` pins already used elsewhere in the repo. We're consolidating workflow-hygiene changes in #4103 (the PR that restructures these jobs into build-once + sharding) rather than duplicating them across the stack; #4102 merges immediately before it.

File: .github/workflows/pr-checks.yml [hunk 3]
```diff
@@ -92,6 +92,42 @@ jobs:
         path: test-results/
         retention-days: 7
 
+  accessibility-tests:
+    name: Accessibility Tests
+    runs-on: ubuntu-22.04
+    # Backstop, as on `runtime-tests` — same ~440-test/retry profile, plus a
+    # slower per-test axe scan, so a little more headroom.
+    timeout-minutes: 45
```
[the remark this answers is not in the document]
Comment (adrianschmidt) [reply .github/workflows/pr-checks.yml:100]: @coderabbitai good point. Handled in the stacked #4103 (commit 36b4183f5), which adds `permissions: { contents: read }` to the example-test jobs (`build-example-docs`, `runtime-tests`, `accessibility-tests`). Scoped to those jobs rather than a workflow-wide default on purpose: a top-level default would also constrain `sonarcloud` (PR decoration via `GITHUB_TOKEN`) and `autosquash`, which rely on the broader default token. Keeping this PR focused on the accessibility feature.

**Review by Kiarokh — APPROVED** [review]
